### PR TITLE
EAMxx: rewrite interfaces to scorpio to bypass completely F90 lib

### LINKED ISSUE #792: [AD] SCORPIO-Interface: Skip the F90 bridge and call the C version of SCORPIO directly
Currently the SCORPIO interface follows the EAM paradigm of using F90 code to call SCORPIO F90 code, which is really a bridge to SCORPIO C code.  Because SCREAM is written in C++ this means we,

Call C++ code, which bridges to F90 code, which calls SCORPIO-F90 code, which bridges to SCORPIO-C code.

This task is to cut out the middle man of F90 code and just go from C++ SCREAM to C SCORPIO.

This task involves a number of sub-tasks and can be broken into bite sized pieces.  The SCREAM to SCORPIO interface described above is made up of a number of individual routines.  Each of which could be ported independently.

NOTE:  This task *does* provide an opportunity to rethink the SCREAM->SCORPIO interface to fit a paradigm where there is no C++ code.  For example, currently, a lot of persistent information is stored in an F90 module.  Removal of F90 means we can redesign how this persistent information is handled, and do so in a C++ way.

In other words, this task may be best served by a pre-planning stage before any implementation is executed.

### PR DESCRIPTION
**Pull request by bartgol**
List of changes:

* Do not use piof (keep the target, b/c Homme needs it for RRM)
* Split interface to pioc into small lib (does not need scream_share). @brhillman should rrtmgp use this instead of simple_netcdf.hpp?
* Add unit tests for interface to pioc
* Changed a few names for interfaces
* Switch to a temporary branch for scorpio, since we need a patch that did not make the cut for the v1.4.1 tag. As soon as a new scorpio release comes out, we can switch back to tags.

This PR avoids jumps to/from Fortran when we use pio, and simplifies handling of lists/memory, by relying on std stuff.

Closes #792

### CODE REVIEW

**Comment by PeterCaldwell** [comment]
Does this completely remove all fortran from our I/O, or just a piece of it? I was under the impression that removing all the fortran would be a huge undertaking...

**Comment by bartgol** [comment]
> Does this completely remove all fortran from our I/O, or just a piece of it? I was under the impression that removing all the fortran would be a huge undertaking...

It removed piof completely from EAMxx itself, but Homme still needs it (for RRM, IIRC).

It was not a huge undertaking after all. I think I did the bulk of it in a day, and then a few tidbits here and there in my spare time at night.

File: components/eamxx/src/share/grid/remap/coarsening_remapper.cpp [hunk 1]
```diff
@@ -55,21 +55,22 @@ CoarseningRemapper (const grid_ptr_type& src_grid,
...
-  scorpio::set_dof(map_file,"row",nlweights,dofs_offsets.data());
-  scorpio::set_dof(map_file,"col",nlweights,dofs_offsets.data());
-  scorpio::set_dof(map_file,"S",nlweights,dofs_offsets.data());
-  scorpio::set_decomp(map_file);
-  scorpio::grid_read_data_array(map_file,"row",-1,row_gids_h.data(),nlweights);
-  scorpio::grid_read_data_array(map_file,"col",-1,col_gids_h.data(),nlweights);
-  scorpio::grid_read_data_array(map_file,"S",  -1,S_h.data(),       nlweights);
-  scorpio::eam_pio_closefile(map_file);
+  for (const std::string& vname : {"row", "col", "S"}) {
+    EKAT_REQUIRE_MSG (scorpio::has_variable(map_file,"row"),
```
Comment (AaronDonahue) [inline components/eamxx/src/share/grid/remap/coarsening_remapper.cpp:61]: Same comment here as elsewhere about `"row"` vs. `name`

File: components/eamxx/src/share/grid/remap/horizontal_remap_utility.cpp [hunk 2]
```diff
@@ -43,7 +43,12 @@ void HorizontalMap::set_remap_segments_from_file(const std::string& remap_filena
   start_timer("EAMxx::HorizontalMap::set_remap_segments_from_file");
   // Open remap file and determine the amount of data to be read
   scorpio::register_file(remap_filename,scorpio::Read);
-  const auto remap_size = scorpio::get_dimlen(remap_filename,"n_s"); // Note, here we assume a standard format of col, row, S
+  for (const std::string& vname : {"row", "col", "S"}) {
+    EKAT_REQUIRE_MSG (scorpio::has_variable(remap_filename,"row"),
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/share/grid/remap/horizontal_remap_utility.cpp:47]: yes

File: components/eamxx/src/control/atmosphere_driver.cpp [hunk 3]
```diff
@@ -777,16 +776,15 @@ void AtmosphereDriver::restart_model ()
...
   m_current_ts.set_num_steps(nsteps);
   m_run_t0.set_num_steps(nsteps);
 
   for (auto& it : m_atm_process_group->get_restart_extra_data()) {
     const auto& name = it.first;
           auto& any  = it.second;
 
-
-    auto data = scorpio::get_any_attribute(filename,name);
+    auto data = scorpio::get_any_attribute(filename,"GLOBAL",name);
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/control/atmosphere_driver.cpp:787]: `get_any_attribyte(file,"GLOBAL",name)` returns an `ekat::any` object, while `get_global_attribute` returns the actual value (provided that the template type is the correct one, of course: you can't call `get_global_attribute<int>(file,att)` if the actual value is a double).

And yes, the methods with `_global_` return global nc attributes, while the others return attributes of the variables.

**Review by AaronDonahue — DISMISSED** [review]
This is a pretty epic PR.  Very happy to be moving away from using an F90 bridge to handle IO.  Thanks @bartgol

**Review by AaronDonahue — DISMISSED** [review]

**Comment by AaronDonahue** [comment]
@bartgol , I approved the PR after your nano-fix.  I noticed it doesn't have the AUTOMERGE label, feel free to add it or manually merge if you think this is ready.

**Comment by bartgol** [comment]
> @bartgol , I approved the PR after your nano-fix. I noticed it doesn't have the AUTOMERGE label, feel free to add it or manually merge if you think this is ready.

I don't want automerge. Once all tests pass, I want to check that everything runs as expected on other machines.

**Comment by bartgol** [comment]
This PR has fallen way out of sync with master. I created a separate branch which incorporates the changes, and it up to date with master. Closing this PR.

**Pull request closed without merging**